### PR TITLE
WIP: ETL summary report basic features

### PR DESCRIPTION
**Pull request by atatmaja**
## Briefly describe the feature or bug that this PR addresses.

Sends a summary report detailing each ETL run to slack. For now we've included the following basic functionality:
- Date & time ETL ran
- ETL runtime
- Record count in each table before and after 
- Record count from Airtable

## Please link the Airtable ticket associated with this PR.

https://airtable.com/tbli2lWQHAqBa6ZcI/viw8iBsG4yooIYvRI/recaU6vCNWIszwnCI

## Describe the steps you took to test the feature/bugfix introduced by this PR.

- Tested locally, made sure that correct summary reports were sent to slack
- Specifically, tested that the summary report would be generated even if an error was raised midway through the ETL, causing it to abort

## Does any infrastructure work need to be done before this PR can be pushed to production?

N/A

### CODE REVIEW

File: app/database_etl/summary_report_generator/summary_report.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+class SummaryReport:
+    def __init__(self):
+        self.start_time = time()
+        self.get_table_counts_before()
+
+    def get_elapsed_time(self):
+        return time() - self.start_time
+
+    def get_num_airtable_records(self, num_records: int):
```
Comment (simonarocco) [inline app/database_etl/summary_report_generator/summary_report.py:18]: shouldn't this be set_num_airtable_records?

File: app/database_etl/summary_report_generator/summary_report.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                table_counts[table.__tablename__] = session.query(func.count(getattr(table, pk))).scalar()
+        return table_counts
+
+    def get_table_counts_before(self):
+        self.table_counts_before = self.get_table_counts()
+
+    def get_table_counts_after(self):
+        self.table_counts_after = self.get_table_counts()
+
+    def get_human_readable_time(self, timestamp: float):
```
Comment (simonarocco) [inline app/database_etl/summary_report_generator/summary_report.py:39]: i feel like we dont rly need this func cause it's just used once and we can put the logic where it's used

File: app/database_etl/summary_report_generator/summary_report.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            body += f"Status: SUCCESS"
+        else:
+            body += f"Status: FAIL"
+        send_slack_message(body, channel="#dev-logging-etl")
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exception_type, exception_value, traceback):
+        self.get_elapsed_time()
```
Comment (simonarocco) [inline app/database_etl/summary_report_generator/summary_report.py:60]: why is this needed? i dont think this function call is needed for getting the elapsed time?

File: app/database_etl/summary_report_generator/summary_report.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    def get_table_counts(self):
+        table_counts = {}
+        with db_session() as session:
+            for table in [DashboardSource, ResearchSource, Country, State, City,
+                          TestManufacturer, AntibodyTarget, CityBridge, StateBridge,
+                          TestManufacturerBridge, AntibodyTargetBridge]:
+                # Inspect primary key of table
+                pk = inspect(table).primary_key[0].key
+                # Use primary key to get row count instead of using the .count() method for perf reasons
+                table_counts[table.__tablename__] = session.query(func.count(getattr(table, pk))).scalar()
```
Comment (simonarocco) [inline app/database_etl/summary_report_generator/summary_report.py:30]: does this work instead? `select([func.count()]).select_from(table)`? reference: https://stackoverflow.com/questions/12941416/how-to-count-rows-with-select-count-with-sqlalchemy

File: app/database_etl/summary_report_generator/summary_report.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        return table_counts
+
+    def get_table_counts_before(self):
```
Comment (simonarocco) [inline app/database_etl/summary_report_generator/summary_report.py:33]: these two functions should be set_

File: app/database_etl/summary_report_generator/summary_report.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        self.table_counts_after = self.get_table_counts()
+
+    def get_human_readable_time(self, timestamp: float):
+        dt = datetime.datetime.fromtimestamp(timestamp)
+        return dt.strftime('%Y-%m-%d %H:%M:%S')
+
+    def send_summary_report(self):
+        body = f"Summary report for ETL run at {self.get_human_readable_time(self.start_time)}:\n\n"
+        body += f"Total runtime: {self.get_elapsed_time()} seconds\n"
+        if hasattr(self, 'num_airtable_records'):
```
Comment (simonarocco) [inline app/database_etl/summary_report_generator/summary_report.py:46]: why would this attribute not get set? it gets set in the ETL no matter what (i.e. not based on a certain condition)

File: app/database_etl/etl_main.py [hunk 7]
```diff
@@ -24,84 +25,89 @@
...
-
-    # Make sure that filter options are still valid
-    check_filter_options(dashboard_source)
-
-    # Upload tableau csv to google sheets
-    upload_analyze_csv()
-    return
+    # etl_report.__exit__() will run whether or not this block is successfully completed
+    # see https://www.geeksforgeeks.org/with-statement-in-python/
+    with SummaryReport() as etl_report:
```
Comment (simonarocco) [inline app/database_etl/etl_main.py:30]: is there a benefit to making this a generator? is it that the summary report will automatically be sent upon cleanup so we don't have to call the send summary report function?

**Review by simonarocco — COMMENTED** [review]
looks good!! thanks for this. some overall questions and comments.

File: app/database_etl/summary_report_generator/summary_report.py (hunk 1 above)
Comment (atatmaja) [reply app/database_etl/summary_report_generator/summary_report.py:18]: yes agree

File: app/database_etl/summary_report_generator/summary_report.py (hunk 2 above)
Comment (atatmaja) [reply app/database_etl/summary_report_generator/summary_report.py:39]: down for abstraction, jk ya sure

File: app/database_etl/summary_report_generator/summary_report.py (hunk 6 above)
Comment (atatmaja) [reply app/database_etl/summary_report_generator/summary_report.py:46]: what if the etl fails before we are able to get a number of airtable records (e.g. if the request to airtable fails)

File: app/database_etl/etl_main.py (hunk 7 above)
Comment (atatmaja) [reply app/database_etl/etl_main.py:30]: yes, basically want to make it so that the send summary report function will always run even if we run into an error midway thru the code

File: app/database_etl/summary_report_generator/summary_report.py (hunk 5 above)
Comment (atatmaja) [reply app/database_etl/summary_report_generator/summary_report.py:33]: agree

File: app/database_etl/summary_report_generator/summary_report.py (hunk 4 above)
Comment (atatmaja) [reply app/database_etl/summary_report_generator/summary_report.py:30]: it does but i am more down to use ORM layer, more consistent with the rest of our sqlalchemy usage

File: app/database_etl/summary_report_generator/summary_report.py (hunk 3 above)
Comment (atatmaja) [reply app/database_etl/summary_report_generator/summary_report.py:60]: u r correct

File: app/database_etl/summary_report_generator/summary_report.py (hunk 2 above)
Comment (atatmaja) [reply app/database_etl/summary_report_generator/summary_report.py:39]: actually I am going to leave this as a function here, it makes the code in send_summary_report look cleaner

**Review by simonarocco — APPROVED** [review]
gg
